### PR TITLE
gRPC health server sets serving status to NOT_SERVING on defrag

### PR DESCRIPTION
**Pull request by chaochn47**
Trying to fix https://github.com/kubernetes/kubernetes/issues/93280 on etcd side since etcd is part of the kubernetes sigs. 

Before failure rate is around 33%
```
v3_failover_test.go:101: etcd client failed to fail over, error (context deadline exceeded)
v3_failover_test.go:103: request failure rate is 33.59%, traffic volume success 172 requests, total 259 requests
```

After failure rate is 0%
```
v3_failover_test.go:103: request failure rate is 0.00%, traffic volume success 2116 requests, total 2116 requests
```

```
2023-07-20T22:18:20.243Z [[core] [Channel #19 SubChannel #20] Subchannel picks a new address "unix:localhost:m0" to connect]
2023-07-20T22:18:20.243Z [[balancer] base.baseBalancer: handle SubConn state change: 0xc0006aad68, CONNECTING]
2023-07-20T22:18:20.243Z [[core] [Channel #19 SubChannel #20] Subchannel Connectivity change to READY]
2023-07-20T22:18:20.243Z [[balancer] base.baseBalancer: handle SubConn state change: 0xc0006aad68, READY]
2023-07-20T22:18:20.243Z [[roundrobin] roundrobinPicker: Build called with info: {map[0xc0006aad68:{{
          "Addr": "unix:localhost:m0",
          "ServerName": "localhost",
          "Attributes": null,
          "BalancerAttributes": null,
          "Type": 0,
          "Metadata": null
        }} 0xc0006aadf8:{{
          "Addr": "unix:localhost:m2",
          "ServerName": "localhost",
          "Attributes": null,
          "BalancerAttributes": null,
          "Type": 0,
          "Metadata": null
        }}]}]


2023-07-20T22:18:20.743Z INFO	m0	starting defragment	{"member": "m0"}
2023-07-20T22:18:20.743Z [[core] [Channel #19 SubChannel #20] Subchannel Connectivity change to TRANSIENT_FAILURE]
2023-07-20T22:18:20.743Z [[balancer] base.baseBalancer: handle SubConn state change: 0xc0006aad68, TRANSIENT_FAILURE]
2023-07-20T22:18:20.743Z [[roundrobin] roundrobinPicker: Build called with info: {map[0xc0006aadb0:{{
          "Addr": "unix:localhost:m1",
          "ServerName": "localhost",
          "Attributes": null,
          "BalancerAttributes": null,
          "Type": 0,
          "Metadata": null
        }} 0xc0006aadf8:{{
          "Addr": "unix:localhost:m2",
          "ServerName": "localhost",
          "Attributes": null,
          "BalancerAttributes": null,
          "Type": 0,
          "Metadata": null
        }}]}]
```

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

**Comment by chaochn47** [comment]
Hi @ahrtr @serathius @jmhbnz @fuweid, could you please take a look this approach and share your thoughts about it? 

> 1. making defrag not blocking quorum reads
> 2. making etcd client's load balancer switch to a follower if the leader is doing defrag. (I believe today we only have round robin LB, am I correct?)

I was inspired by what https://github.com/kubernetes/kubernetes/issues/93280#issuecomment-664013380 proposed in the k8s issue. 

It will be helpful for eliminating online defrag impact. 

/cc @lavacat @dims

File: server/etcdserver/api/v3rpc/maintenance.go
Comment (chaochn47) [inline server/etcdserver/api/v3rpc/maintenance.go:92]: There are chances that with a write-intensive workload, it will take some time for the local node to apply the backlogged in-flights entries. 

But this PR should make etcd availability much better when defrag is active.

**Comment by lavacat** [comment]
> Please see [#16276 (comment)](https://github.com/etcd-io/etcd/issues/16276#issuecomment-1651845511)
@ahrtr are you referring to Livez/Readyz, 3.7 defrag plans or experimental dependency? 
 - Livez/Readyz: It will be a larger change and imo will build on this implementation. Also, current [design](https://docs.google.com/document/d/1PaUAp76j1X92h3jZF47m32oVlR8Y-p-arB5XOB7Nb6U/edit) doesn't address client side traffic failover during defrag.
 - 3.7 defrag: not clear if it will be back ported to 3.5 and will take longer time to implement.
 -  experimental dependency: I don't see this PR adding new dependencies. It does add `"healthCheckConfig": {"serviceName": ""}}` deepening current dependency (adds new [stream](https://github.com/grpc/grpc-go/blob/v1.58.2/health/client.go#L59C10-L59C10) behind the scenes).

IMO this PR is a bugfix for an existing problem with defrag. Fundamentally if we don't want to use https://pkg.go.dev/google.golang.org/grpc/health, we shouldn't merge this,  otherwise I think it's very useful fix.

File: tests/e2e/failover_test.go
Comment (chaochn47) [reply tests/e2e/failover_test.go:88]: etcd client flag can be added separately as a follow up. 

I am trying to limit the required changes so
1. we could back port the change in 3.5 internally first
2. apiserver just need to add a patch without waiting for the etcd 3.5 release

**Comment by fuweid** [comment]
@chaochn47 For the change, it looks good to me. It's helpful to the kube-apiserver which has multiple endpoints instead of one service. The kube-apiserver can forward the requests to available servers.  What do you think about introducing the flag to enable this feature? It might bring heavy traffic to other servers.

**Comment by chaochn47** [comment]
Thanks @fuweid for the feedback! 

> What do you think about introducing the flag to enable this feature? It might bring heavy traffic to other servers.
Yeah, the heuristic is defrag won't take too long and once it's completed. New requests will be routed to this server and watch will break and has 1/3 chance to establish on this server. 

There is no built-in supported capacity aware load balancer in gRPC-go. I will explore and explain it in the [design doc](https://docs.google.com/document/d/1x7guECGQI1wdAh2fBzPBRTva-sqGGLR8Bp0lrrmmDk0/edit#heading=h.euv7j5s250nm) after livez/readyz.

**Review by lavacat — APPROVED** [review]

File: server/etcdserver/api/v3rpc/health.go
Comment (ahrtr) [inline server/etcdserver/api/v3rpc/health.go:45]: Let's ensure `lg` will never be `nil`. Same below.
```suggestion
		hc.lg.Info("start serving gRPC requests from client")
```

File: server/etcdserver/api/v3rpc/health.go
Comment (ahrtr) [inline server/etcdserver/api/v3rpc/health.go:29]:
```suggestion
        if lg == nil {
		lg = zap.NewNop()
	}
	hc := &healthChecker{hs: hs, lg: lg}
```

File: tests/e2e/failover_test.go
Comment (ahrtr) [inline tests/e2e/failover_test.go:127]:
```suggestion
	err := clus.Procs[0].Failpoints().Setup(context.Background(), "defragBeforeCopy", `sleep("8s")`)
```

File: tests/e2e/failover_test.go
Comment (ahrtr) [inline tests/e2e/failover_test.go:56]: only one case in the table? Are you going to add more in followup PR?

File: tests/e2e/failover_test.go
Comment (ahrtr) [inline tests/e2e/failover_test.go:107]: In exactly what cases, the request may fail?

**Comment by ahrtr** [comment]
Please also rebase the PR.

File: tests/e2e/failover_test.go
Comment (chaochn47) [reply tests/e2e/failover_test.go:56]: Potentially, I would imagine gRPC health server will re-use the readyz checks.

File: tests/e2e/failover_test.go
Comment (chaochn47) [reply tests/e2e/failover_test.go:107]: > Because of the inherently asynchronous nature of the network, whenever a backend transitions from healthy to unhealthy, it may still receive some number of RPCs that were already in flight from the client before the client received the notification that the backend is unhealthy. This race condition lasts approximately the network round trip time (it includes the one-way trip time for RPCs that were already sent by the client but not yet received by the server when the server goes unhealthy, plus the one-way trip time for RPCs sent by the client between when the server sends the unhealthy notification and when the client receives it).

https://github.com/grpc/proposal/blob/master/A17-client-side-health-checking.md#race-conditions-and-subchannel-startup

RTT should be around 10ms if client and server are deployed in different host but in the same region.

**Comment by chaochn47** [comment]
> Please also rebase the PR.

Done. 

---

IMHO, the PR itself is good to merge and safe to cherry picked to release branch since it's recommended to do defrag against one member at a time. 

One of the remaining risks to evaluate before expanding this to other readyz checks is client should fail open and at least keep one connection in the picker if non of the etcd health servers is SERVING.

File: tests/e2e/failover_test.go
Comment (ahrtr) [reply tests/e2e/failover_test.go]: Then pls leave a TODO comment here, thx

**Comment by ahrtr** [comment]
I see that the e2e test case is very similar to the integration test case, is it possible to write a common test case?

**Comment by chaochn47** [comment]
> I see that the e2e test case is very similar to the integration test case, is it possible to write a common test case?

Enabling failpoint both in e2e and integration test has not yet built into the common test framework. This can be a separate PR to follow up. 

> Another big question is should we add an experimental flag something like "--experimental-stop-grpc-service-on-defragmentation"? I think the answer should be YES. Any new feature might also bring new problem, at least we add an option for users to disable it. 

Okay.

**Comment by chaochn47** [comment]
Ping @ahrtr

File: server/config/config.go
Comment (ahrtr) [inline server/config/config.go:195]:
```suggestion
	// ExperimentalStopGRPCServiceOnDefrag enables etcd gRPC service to stop serving client requests on defragmentation.
```

File: server/embed/config.go
Comment (ahrtr) [inline server/embed/config.go:432]:
```suggestion
	// ExperimentalStopGRPCServiceOnDefrag enables etcd gRPC service to stop serving client requests on defragmentation.
```

File: server/embed/config.go
Comment (ahrtr) [inline server/embed/config.go:732]:
```suggestion
	fs.BoolVar(&cfg.ExperimentalStopGRPCServiceOnDefrag, "experimental-stop-grpc-service-on-defrag", cfg.ExperimentalStopGRPCServiceOnDefrag, "Enable etcd gRPC service to stop serving client requests on defragmentation.")
```

File: server/etcdmain/help.go
Comment (ahrtr) [inline server/etcdmain/help.go:309]:
```suggestion
    Enable etcd gRPC service to stop serving client requests on defragmentation.
```

File: tests/integration/v3_failover_test.go
Comment (ahrtr) [inline tests/integration/v3_failover_test.go:190]:
```suggestion
	integration2.BeforeTest(t, integration2.WithFailpoint("defragBeforeCopy", `sleep(10s)`))
```

File: server/config/config.go
Comment (ahrtr) [inline server/config/config.go:195]:
```suggestion
	// ExperimentalStopGRPCServiceOnDefrag enables etcd gRPC service to stop serving client requests on defragmentation.
```

**Review by ahrtr — APPROVED** [review]
LGTM

Good job! thx @chaochn47

**Comment by chaochn47** [comment]
Ping @serathius

File: server/etcdserver/api/v3rpc/health.go
Comment (serathius) [inline server/etcdserver/api/v3rpc/health.go:45]: Why would hc.hs be ever nil? Let's avoid handling cases that should not happen during correct code execution.

File: server/etcdserver/api/v3rpc/health.go
Comment (serathius) [inline server/etcdserver/api/v3rpc/health.go:50]: Can you obfuscate the value behind "". For example create const called `allGRPCServices` and use it here.

File: tests/e2e/failover_test.go
Comment (serathius) [inline tests/e2e/failover_test.go:108]: Please don't use short names.  Especially if the variable is long lived.

File: tests/integration/v3_failover_test.go
Comment (serathius) [inline tests/integration/v3_failover_test.go:42]: Please avoid using comments to suplement the name. Just call it `expectedMaxFailedRequests`.

File: tests/integration/v3_failover_test.go
Comment (serathius) [inline tests/integration/v3_failover_test.go:209]: Please create a separate defer for each variable as they are used.

File: tests/integration/v3_failover_test.go
Comment (serathius) [inline tests/integration/v3_failover_test.go:196]: Please use `sync.WorkGroup`

File: tests/integration/v3_failover_test.go
Comment (serathius) [inline tests/integration/v3_failover_test.go:33]: No need to implement both integration and e2e tests. E2e tests is enough.

File: tests/e2e/failover_test.go
Comment (serathius) [inline tests/e2e/failover_test.go:51]: Recommend always mirroring a test. Testing only feature enabled can be really fragile, we cannot distinquish whether reason for test passing is the new feature or there was some external change.

Recommend to add a test case with `WithExperimentalStopGRPCServiceOnDefrag(false)` that is expected to fail requests. Instead of checking maximum number of failed requests, there you can check the minimum number of failed requests.

File: tests/e2e/failover_test.go
Comment (serathius) [inline tests/e2e/failover_test.go:77]: Pre-defining a `cc` before it's instantiated and then checking for nil is not very readable.

Please create a separate defer for each variable as they are used.

File: tests/e2e/failover_test.go
Comment (serathius) [inline tests/e2e/failover_test.go:42]: Please avoid using comments to suplement the name. Just call it `expectedMaxFailedRequests`.

File: tests/e2e/failover_test.go
Comment (serathius) [inline tests/e2e/failover_test.go:72]: Please use `sync.WaitGroup` to ensure goroutine exits.

Comment (serathius) [reply tests/e2e/failover_test.go:72]: Or even better `errgroup.Group` as it allows to return an error which would remove the need for `errc`.

File: tests/e2e/failover_test.go
Comment (serathius) [inline tests/e2e/failover_test.go:126]: Note, you should also check minimal number of requests executed. Environment could allow for only 5 requests passing. To make sure the test is working as intended we should require at least 20 or more requests.

File: server/etcdserver/api/v3rpc/health.go
Comment (chaochn47) [reply server/etcdserver/api/v3rpc/health.go:45]: okay. Switched to assert with panic result in that case.

File: server/etcdserver/api/v3rpc/health.go
Comment (chaochn47) [reply server/etcdserver/api/v3rpc/health.go]: Done

File: tests/e2e/failover_test.go
Comment (chaochn47) [reply tests/e2e/failover_test.go]: Done

File: tests/integration/v3_failover_test.go
Comment (chaochn47) [reply tests/integration/v3_failover_test.go]: Removed integration test.

File: tests/e2e/failover_test.go
Comment (chaochn47) [reply tests/e2e/failover_test.go:60]: Implemented negative tests.

File: tests/e2e/failover_test.go
Comment (chaochn47) [reply tests/e2e/failover_test.go]: Done

File: tests/e2e/failover_test.go
Comment (chaochn47) [reply tests/e2e/failover_test.go]: Done

File: tests/e2e/failover_test.go
Comment (chaochn47) [reply tests/e2e/failover_test.go]: Done

File: tests/e2e/failover_test.go
Comment (chaochn47) [reply tests/e2e/failover_test.go]: Done

**Comment by chaochn47** [comment]
Thanks @serathius for the review. I have added aggressive test threshold for happy and negative test cases.

**Comment by chaochn47** [comment]
E2E test failed in 386 arch. Reason is `gofail` is not enabled in CI. Test should be skipped after rebasing on top of the changes https://github.com/etcd-io/etcd/pull/16698 introduces.

```
2023-10-18T19:24:18.1851440Z     failover_test.go:145: request failure rate is 0.00%, traffic volume successfulRequestCount 15786 requests, total 15786 requests
2023-10-18T19:24:18.1852640Z     failover_test.go:169: 
2023-10-18T19:24:18.1853387Z         	Error Trace:	/__w/etcd/etcd/tests/e2e/failover_test.go:169
2023-10-18T19:24:18.1854532Z         	            				/__w/etcd/etcd/tests/e2e/failover_test.go:146
2023-10-18T19:24:18.1855364Z         	Error:      	"0" is not greater than or equal to "90"
2023-10-18T19:24:18.1857108Z         	Test:       	TestFailoverOnDefrag/defrag_blocks_one-third_of_requests_with_stopGRPCServiceOnDefrag_set_to_true_and_client_health_check_disabled
```

ref. https://github.com/etcd-io/etcd/actions/runs/6565503688/job/17834165379?pr=16278

File: server/etcdserver/api/v3rpc/health.go
Comment (serathius) [inline server/etcdserver/api/v3rpc/health.go:49]: If you are worried about hc.hs being nil, why not check in the `NewHealthNotifier` ?

File: tests/e2e/failover_test.go
Comment (serathius) [inline tests/e2e/failover_test.go:58]: Please avoid a dedicated functions for assertions, instead have two variables, `expectedMinTotalRequestsCount`, `expectedMaxFailedRequestsCount` set on tcs and inline the assetions into TestFailoverOnDefrag function

```
	if tc.expectedMinTotalRequestsCount {
		require.GreaterOrEqual(t, requestVolume, tc.expectedMinTotalRequestsCount)
	}
	if tc.expectedMaxFailedRequestsCount {
		failedRequestCount := requestVolume - successfulRequestCount
		require.LessOrEqual(t, failedRequestCount, expectedMaxFailedRequestsCount)
	}
```

File: tests/e2e/failover_test.go
Comment (serathius) [inline tests/e2e/failover_test.go:108]: Can you make `grpcDialOptions` a field on `tc`. Having it directly on test case will make it more readable when reading the test cases vs some custom variable that needs to be looked up.

Comment (chaochn47) [reply tests/e2e/failover_test.go]: Make sense to me. Done.

File: tests/e2e/failover_test.go
Comment (chaochn47) [reply tests/e2e/failover_test.go]: Done!

File: server/etcdserver/api/v3rpc/health.go
Comment (chaochn47) [reply server/etcdserver/api/v3rpc/health.go]: Good idea :p Done

**Comment by chaochn47** [comment]
Kindly ping @serathius @wenjiaswe @ptabor @mitake @spzala @jmhbnz ~

**Comment by chaochn47** [comment]
Hi @ahrtr @fuweid 

What's the suggested process to move progress on this PR? I think all the review comments are addressed.

**Review by ahrtr — APPROVED** [review]

**Comment by ahrtr** [comment]
This is a nice to have feature, it can prevent client from being blocked on the member in progress of defragmentation.

We also have flag `--experimental-stop-grpc-service-on-defrag` (defaults to `false`). There is no any impact on users if they don't intentionally enable it. So it should be safe.

Leave to other maintainers to take a second look. 

Thanks @chaochn47

File: server/etcdserver/api/v3rpc/health.go
Comment (serathius) [inline server/etcdserver/api/v3rpc/health.go:42]: Nit: Calling `StartServe()` here means that etcd will emit `"start serving gRPC requests from client"` log which is incorrect. `NewHealthNotifier` doesn't control whether grpc is serving, this is done in different place. Health notifier only changes the serving status. I would change the log line to 
```
hc.lg.Info("grpc service status changed", zap.String("service", allGRPCServices), zap.String("status", "serving")) 
```

**Review by serathius — APPROVED** [review]

File: server/etcdserver/api/v3rpc/health.go
Comment (chaochn47) [reply server/etcdserver/api/v3rpc/health.go:42]: Makes sense. Updated the logging.

File: server/etcdserver/api/v3rpc/health.go
Comment (chaochn47) [reply server/etcdserver/api/v3rpc/health.go]: Duplicated comments are removed now.

File: tests/e2e/failover_test.go
Comment (serathius) [inline tests/e2e/failover_test.go:56]: Idea for followup, better to assert on rate then on request count. Minimal request count is good, but what happens if one machines serves 300 requests, while other 3000? 90 failures out of 300 is 30% while only 3 % on 3000.

File: server/etcdserver/api/v3rpc/health.go
Comment (serathius) [inline server/etcdserver/api/v3rpc/health.go:27]: Idea for followup. Instead of allowing defrag to change StartServe status as they will, better would be to just notify that defrag is happening and let health notify decide on health status. https://en.wikipedia.org/wiki/Law_of_Demeter, why defrag should know how about changing health.

Comment (serathius) [reply server/etcdserver/api/v3rpc/health.go:27]: Additional improvement would be that Maintenance server would no longer need to be aware of `stopServingOnDefrag` flag, instead configuration would be handled by health.

**Comment by serathius** [comment]
I think this is good enough to merge, still please consider implementing followups.

File: server/etcdserver/api/v3rpc/health.go
Comment (chaochn47) [reply server/etcdserver/api/v3rpc/health.go:27]: Thanks! I like the idea of decoupling stop serving from the maintenance server. 

Eventually the health notifier may extend to other readyz checks (it need more in-depth thought though) implemented in the readyz handler.

File: tests/e2e/failover_test.go
Comment (chaochn47) [reply tests/e2e/failover_test.go:56]: Gotcha, something like the robustness TPS assertion.

**Pull request merged**
